src/Orchard/Security/Providers/FormsAuthenticationService.cs

```diff
@@ -135,39 +109,114 @@ public class FormsAuthenticationService : IAuthenticationService {
             }
 
             var formsIdentity = (FormsIdentity)httpContext.User.Identity;
-            var userData = formsIdentity.Ticket.UserData ?? "";
 
-            // The cookie user data is {userName.Base64};{tenant}.
-            var userDataSegments = userData.Split(';');
-
-            if (userDataSegments.Length < 2) {
-                return null;
-            }
-
-            var userDataName = userDataSegments[0];
-            var userDataTenant = userDataSegments[1];
-
-            try {
-                userDataName = userDataName.FromBase64();
+            var userData = formsIdentity.Ticket.UserData ?? "";
+            var userDataDictionary = new Dictionary<string, string>();
+            
+            if (formsIdentity.Ticket.Version == 3) {
+                var userDataSegments = userData.Split(';');
```

Comment:
Just revoke the cookie.

Reply:
It's what I initially thought of doing. However that would have meant that deploying this update to how the cookies are processed would have caused all users to effectively be logged out. 
That would have been jarring in terms of UX, and no LoggedOut event would have been fired for anyone.

Something else I thought would be adding at the end of this method a condition to replace V3 cookies with new ones by releasing it (I'd just call CreateAndAddCookie)

Reply:
I like the idea of replacing as a good compromise.